File: Fourth Semester - Design and Programming Environments/Java/ProiectJava/java/org/example/repository/impl/ReservationRepositoryImpl.java
package org.example.repository.impl;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.example.exception.RepositoryException;
import org.example.modul.Reservation;
import org.example.modul.Trip;
import org.example.modul.validator.Validator;
import org.example.repository.interfaces.ReservationRepository;
import org.example.repository.utils.JdbcUtils;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.Properties;

public class ReservationRepositoryImpl implements ReservationRepository {
    protected JdbcUtils jdbcUtils;
    protected Validator<Reservation> validator;
    protected static final Logger logger = LogManager.getLogger();

    public ReservationRepositoryImpl(Properties properties, Validator<Reservation> validator) {
        logger.info("Initializing ReservationDatabaseRepository with properties: {}", properties);

        this.jdbcUtils = new JdbcUtils(properties);
        this.validator = validator;
    }

    private Reservation getReservationFromSet(ResultSet result) throws SQLException {
        Integer id = result.getInt("id");
        Integer numberOfReservedSeats = result.getInt("numberOfReservedSeats");

        String nameClient = result.getString("clientName");
        String contactClient = result.getString("clientContact");

        Integer tripId = result.getInt("tripId");
        String destination = result.getString("destination");
        String transportCompany = result.getString("transportCompany");
        String departure = result.getString("departureDate");
        Double price = result.getDouble("price");
        Integer numberOfAvailableSeats = result.getInt("numberOfAvailableSeats");
        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm");
        LocalDateTime departureDate = LocalDateTime.parse(departure, formatter);
        Trip trip = new Trip(destination, transportCompany, departureDate, price, numberOfAvailableSeats);
        trip.setId(tripId);

        Reservation reservation = new Reservation(trip, nameClient, contactClient, numberOfReservedSeats);
        reservation.setId(id);
        return reservation;
    }

    @Override
    public Optional<Reservation> findOne(Integer id) {
        //how manage if trip, client or reservation is null/

        logger.traceEntry("finding reservation with id {}", id);

        Connection connection = jdbcUtils.getConnection();

        String sql = "select " +
                "r.id as id, r.numberOfReservedSeats as numberOfReservedSeats, " +
                "r.clientName AS clientName, r.clientContact AS clientContact, " +
                "t.id as tripId, t.destination as destination, t.transportCompany as transportCompany, " +
                "t.departureDate AS departureDate, t.price AS price, t.numberOfAvailableSeats as numberOfAvailableSeats " +
                "from reservation r " +
                "inner join trip t on t.id = r.tripId " +
                "where r.id = ?";
        Reservation reservation = null;
        try (PreparedStatement preparedStatement = connection.prepareStatement(sql)) {
            preparedStatement.setInt(1, id);
            try (ResultSet result = preparedStatement.executeQuery()) {
                if (result.next()) {
                    reservation = getReservationFromSet(result);
                    logger.trace("Found reservation: {}", reservation);
                } else {
                    logger.warn("No reservation found with id {}", id);
                }
            }
        } catch (SQLException exception) {
            logger.error(exception);
        }

        logger.traceExit();
        return Optional.ofNullable(reservation);
    }

    @Override
    public Iterable<Reservation> findAll() {
        logger.traceEntry();

        Connection connection = jdbcUtils.getConnection();
        List<Reservation> reservations = new ArrayList<>();

        String sql = "select " +
                "r.id as id, r.numberOfReservedSeats as numberOfReservedSeats, " +
                "r.clientName AS clientName, r.clientContact AS clientContact, " +
                "t.id as tripId, t.destination as destination, t.transportCompany as transportCompany, " +
                "t.departureDate AS departureDate, t.price AS price, t.numberOfAvailableSeats as numberOfAvailableSeats " +
                "from reservation r " +
                "inner join trip t on t.id = r.tripId ";
        try (PreparedStatement preparedStatement = connection.prepareStatement(sql)) {
            try (ResultSet result = preparedStatement.executeQuery()) {
                while (result.next()) {
                    Reservation reservation = getReservationFromSet(result);
                    if (reservation != null) {
                        reservations.add(reservation);
                    }
                }
            }
        } catch (SQLException exception) {
            logger.error(exception);
        }

        logger.traceExit("Found reservations: {}", reservations);
        return reservations;
    }

    @Override
    public Optional<Reservation> save(Reservation entity) {
        logger.traceEntry("saving reservation {}", entity);

        validator.validate(entity);

        Connection connection = jdbcUtils.getConnection();

        try (PreparedStatement preparedStatement = connection.prepareStatement("insert into reservation (tripId, clientName, clientContact, numberOfReservedSeats) values (?, ?, ?, ?)")) {
            preparedStatement.setInt(1, entity.getTrip().getId());
            preparedStatement.setString(2, entity.getClientName());
            preparedStatement.setString(3, entity.getClientContact());
            preparedStatement.setInt(4, entity.getNumberOfReservedSeats());
            int result = preparedStatement.executeUpdate();
            if (result == 0) {
                logger.warn("Failed to save reservation {}", entity);
            } else {
                Optional<Reservation> reservationF = findReservation(entity.getTrip().getId(), entity.getClientName(), entity.getClientContact(), entity.getNumberOfReservedSeats());
                entity.setId(reservationF.get().getId());
                logger.trace("Successfully saved reservation {}", entity);
            }
        } catch (SQLException exception) {
            logger.error(exception);
            throw new RepositoryException("Error saving reservation in database", exception);
        }

        logger.traceExit();
        return Optional.of(entity);
    }

    @Override
    public Optional<Reservation> delete(Integer integer) {
        logger.traceEntry("deleting reservation with id {}", integer);

        if (integer == null || integer < 0) {
            throw new IllegalArgumentException("Reservation id is null or negative");
        }
        Optional<Reservation> reservation = findOne(integer);
        if (reservation.isEmpty()) {
            throw new RepositoryException("This reservation doesn't exists");
        }

        Connection connection = jdbcUtils.getConnection();

        try (PreparedStatement preparedStatement = connection.prepareStatement("delete from reservation where id = ?")) {
            preparedStatement.setInt(1, integer);

            int result = preparedStatement.executeUpdate();
            if (result == 0) {
                logger.warn("Failed to delete reservation {}", reservation);
            } else {
                logger.trace("Successfully deleted reservation {}", reservation);
            }
        } catch (SQLException exception) {
            logger.error(exception);
            throw new RepositoryException("Error deleting reservation in database", exception);
        }

        logger.traceExit();
        return reservation;
    }

    @Override
    public Optional<Reservation> update(Reservation reservation) {
        logger.traceEntry("updating reservation {}", reservation);

        validator.validate(reservation);

        Connection connection = jdbcUtils.getConnection();

        try (PreparedStatement preparedStatement = connection.prepareStatement("update reservation set tripId = ?, " +
                "clientName = ?, clientName = ?, numberOfReservedSeats = ? where id = ?")) {
            preparedStatement.setInt(1, reservation.getTrip().getId());
            preparedStatement.setString(2, reservation.getClientName());
            preparedStatement.setString(3, reservation.getClientContact());
            preparedStatement.setInt(4, reservation.getNumberOfReservedSeats());
            preparedStatement.setInt(5, reservation.getId());
            int result = preparedStatement.executeUpdate();

            if (result == 0) {
                logger.warn("Failed to update reservation {}", reservation);
            } else {
                logger.trace("Successfully updated reservation {}", reservation);
            }
        } catch (SQLException exception) {
            logger.error(exception);
            throw new RepositoryException("Error updating reservation in database", exception);
        }

        logger.traceExit();
        return Optional.of(reservation);
    }

    @Override
    public Optional<Reservation> findReservation(Integer tripId, String nameClient, String contactClient, Integer numberOfReservedSeats) {
        logger.trace("Finding reservation with tripId {}, clientName {}, contactClient {}, numberOfReservedSeats {}",
                tripId, nameClient, contactClient, numberOfReservedSeats);

        String sql = "SELECT " +
                "r.id AS id, r.numberOfReservedSeats AS numberOfReservedSeats, " +
                "r.clientName AS clientName, r.clientContact AS clientContact, " +
                "t.id AS tripId, t.destination AS destination, t.transportCompany AS transportCompany, " +
                "t.departureDate AS departureDate, t.price AS price, t.numberOfAvailableSeats AS numberOfAvailableSeats " +
                "FROM reservation r " +
                "INNER JOIN trip t ON t.id = r.tripId " +
                "WHERE r.tripId = ? AND r.clientName = ? AND r.clientContact = ? AND r.numberOfReservedSeats = ?";

        try (Connection connection = jdbcUtils.getConnection();
             PreparedStatement preparedStatement = connection.prepareStatement(sql)) {

            preparedStatement.setInt(1, tripId);
            preparedStatement.setString(2, nameClient);
            preparedStatement.setString(3, contactClient);
            preparedStatement.setInt(4, numberOfReservedSeats);

            try (ResultSet result = preparedStatement.executeQuery()) {
                if (result.next()) {
                    Reservation reservation = getReservationFromSet(result);
                    logger.trace("Found reservation: {}", reservation);
                    return Optional.of(reservation);
                } else {
                    logger.warn("No reservation found with tripId {}, clientName {}, contactClient {}, numberOfReservedSeats {}",
                            tripId, nameClient, contactClient, numberOfReservedSeats);
                }
            }
        } catch (SQLException e) {
            logger.error("Database error: ", e);
        }

        logger.traceExit();
        return Optional.empty();
    }

    @Override
    public Iterable<Reservation> findReservationsByTrip(Integer tripId) {
        logger.trace("Finding reservations with tripId {}", tripId);

        Connection connection = jdbcUtils.getConnection();
        List<Reservation> reservations = new ArrayList<>();

        String sql = "select " +
                "r.id as id, r.numberOfReservedSeats as numberOfReservedSeats, " +
                "r.clientName AS clientName, r.clientContact AS clientContact, " +
                "t.id as tripId, t.destination as destination, t.transportCompany as transportCompany, " +
                "t.departureDate AS departureDate, t.price AS price, t.numberOfAvailableSeats as numberOfAvailableSeats " +
                "from reservation r " +
                "inner join trip t on t.id = r.tripId " +
                "where r.tripId = ?";
        try (PreparedStatement preparedStatement = connection.prepareStatement(sql)) {
            preparedStatement.setInt(1, tripId);
            try (ResultSet result = preparedStatement.executeQuery()) {
                while (result.next()) {
                    Reservation reservation = getReservationFromSet(result);
                    if (reservation != null) {
                        reservations.add(reservation);
                    }
                }
            }
        } catch (SQLException exception) {
            logger.error(exception);
        }

        logger.traceExit("Found reservations: {}", reservations);
        return reservations;
    }

    @Override
    public Iterable<Reservation> findReservationsByClient(String clientName) {
        logger.trace("Finding reservations with clientName {}", clientName);

        Connection connection = jdbcUtils.getConnection();

        List<Reservation> reservations = new ArrayList<>();
        String sql = "select " +
                "r.id as id, r.numberOfReservedSeats as numberOfReservedSeats, " +
                "r.clientName AS clientName, r.clientContact AS clientContact, " +
                "t.id as tripId, t.destination as destination, t.transportCompany as transportCompany, " +
                "t.departureDate AS departureDate, t.price AS price, t.numberOfAvailableSeats as numberOfAvailableSeats " +
                "from reservation r " +
                "inner join trip t on t.id = r.tripId " +
                "where r.clientName = ?";
        try (PreparedStatement preparedStatement = connection.prepareStatement(sql)) {
            preparedStatement.setString(1, clientName);
            try (ResultSet result = preparedStatement.executeQuery()) {
                while (result.next()) {
                    Reservation reservation = getReservationFromSet(result);
                    reservations.add(reservation);
                }
            }
        } catch (SQLException exception) {
            logger.error(exception);
        }

        logger.traceExit("Found reservations: {}", reservations);
        return reservations;
    }

}

File: Fourth Semester - Design and Programming Environments/C#/ProiectC#/repository/impl/ReservationRepositoryImpl.cs
using TravelAgency.model;
using System.Data;
using log4net;
using TravelAgency.exception;
using TravelAgency.model.validator;

namespace TravelAgency.repository.database;

public class ReservationRepositoryImpl : IReservationRepository
{
    private static readonly ILog log =
        LogManager.GetLogger(System.Reflection.MethodBase.GetCurrentMethod().DeclaringType);

    private IValidator<Reservation>? _reservationValidator;

    private IDictionary<String, String> properties;

    public ReservationRepositoryImpl(IDictionary<String, String> properties, IValidator<Reservation>? reservationValidator)
    {
        log.Info("Creating ReservationDatabaseRepository");

        this.properties = properties;
        this._reservationValidator = reservationValidator;
    }
    
    private Reservation GetReservationFromResult(IDataReader dataReader)
    {
        int id = Convert.ToInt32(dataReader["id"]);
        int numberOfReservedSeats = Convert.ToInt32(dataReader["numberOfReservedSeats"]);
        
        int tripId = Convert.ToInt32(dataReader["tripId"]);
        string destination = dataReader["destination"].ToString();
        string transportCompany = dataReader["transportCompany"].ToString();
        string departureDate = dataReader["departureDate"].ToString();
        double price = Convert.ToDouble(dataReader["price"]);
        int numberOfAvailableSeats = Convert.ToInt32(dataReader["numberOfAvailableSeats"]);
        DateTime departure = DateTime.ParseExact(departureDate, "yyyy-MM-dd HH:mm", System.Globalization.CultureInfo.InvariantCulture);
        Trip? trip = new Trip(destination, transportCompany, departure, price, numberOfAvailableSeats);
        trip.Id = tripId;
        
        string clientName = dataReader["clientName"].ToString();
        string clientContact = dataReader["clientContact"].ToString();
        
        Reservation? reservation = new Reservation(trip, clientName, clientContact, numberOfReservedSeats);
        reservation.Id = id;
        return reservation;
    }
    
    
    public Reservation? FindOne(int id)
    {
        log.Info("Finding Reservation");

        IDbConnection connection = DBUtils.getConnection(properties);

        if (id == null)
        {
            throw new RepositoryException("Reservation Id is null");
        }

        Reservation? reservation = null;
        using (var command = connection.CreateCommand())
        {
            command.CommandText = "select " +
                                  "r.id as id, r.numberOfReservedSeats as numberOfReservedSeats, " +
                                  "r.clientName AS clientName, r.clientContact AS clientContact, " +
                                  "t.id as tripId, t.destination as destination, t.transportCompany as transportCompany, " +
                                  "t.departureDate AS departureDate, t.price AS price, t.numberOfAvailableSeats as numberOfAvailableSeats " +
                                  "from reservation r " +
                                  "inner join trip t on t.id = r.tripId " +
                                  "where r.id = @id";
            IDbDataParameter parameter = command.CreateParameter();
            parameter.ParameterName = "@id";
            parameter.Value = id;

            command.Parameters.Add(parameter);

            using (var dataReader = command.ExecuteReader())
            {
                if (dataReader.Read())
                {
                    reservation = GetReservationFromResult(dataReader);
                    log.InfoFormat("Exiting FindOne Reservation with value: " + reservation);
                    return reservation;
                }
            }
        }

        log.InfoFormat("Exiting FindOne Reservation with value: " + reservation);
        return reservation;
    }

    public IEnumerable<Reservation> FindAll()
    {
        log.Info("Finding All Reservations");
        IDbConnection connection = DBUtils.getConnection(properties);
        IList<Reservation> reservations = new List<Reservation>();

        using (var command = connection.CreateCommand())
        {
            command.CommandText = "select " +
                                  "r.id as id, r.numberOfReservedSeats as numberOfReservedSeats, " +
                                  "r.clientName AS clientName, r.clientContact AS clientContact, " +
                                  "t.id as tripId, t.destination as destination, t.transportCompany as transportCompany, " +
                                  "t.departureDate AS departureDate, t.price AS price, t.numberOfAvailableSeats as numberOfAvailableSeats " +
                                  "from reservation r " +
                                  "inner join trip t on t.id = r.tripId ";

            using (var dataReader = command.ExecuteReader())
            {
                while (dataReader.Read())
                {
                    Reservation reservation = GetReservationFromResult(dataReader);
                    reservations.Add(reservation);
                }
            }
        }

        log.InfoFormat("Exiting FindAll reservations with value: " + reservations);
        return reservations;
    }
    
    public Reservation? FindReservation(int tripId, string clientName, string clientContact, int numberOfReservedSeats)
    {
        log.Info("Finding Reservation by all");
        IDbConnection connection = DBUtils.getConnection(properties);

        if (tripId == null || clientName == null || clientContact == null || numberOfReservedSeats == null || numberOfReservedSeats == 0)
        {
            throw new RepositoryException("Reservation tripId or clientName or clientContact or numberOfReservedSeats are null");
        }
        
        Reservation? reservation = null;
        using (var command = connection.CreateCommand())
        {
            command.CommandText = "select " +
                                  "r.id as id, r.numberOfReservedSeats as numberOfReservedSeats, " +
                                  "r.clientName AS clientName, r.clientContact AS clientContact, " +
                                  "t.id as tripId, t.destination as destination, t.transportCompany as transportCompany, " +
                                  "t.departureDate AS departureDate, t.price AS price, t.numberOfAvailableSeats as numberOfAvailableSeats " +
                                  "from reservation r " +
                                  "inner join trip t on t.id = r.tripId " +
                                  "where r.tripId = @tripId and r.clientName = @clientName and r.clientContact = @clientContact and r.numberOfReservedSeats = @numberOfReservedSeats";
            IDbDataParameter tripParameter = command.CreateParameter();
            tripParameter.ParameterName = "@tripId";
            tripParameter.Value = tripId;
            command.Parameters.Add(tripParameter);
            
            IDbDataParameter clientNameParameter = command.CreateParameter();
            clientNameParameter.ParameterName = "@clientName";
            clientNameParameter.Value = clientName;
            command.Parameters.Add(clientNameParameter);
            
            IDbDataParameter clientContactParameter = command.CreateParameter();
            clientContactParameter.ParameterName = "@clientContact";
            clientContactParameter.Value = clientContact;
            command.Parameters.Add(clientContactParameter);
            
            IDbDataParameter numberOfReservedSeatsParameter = command.CreateParameter();
            numberOfReservedSeatsParameter.ParameterName = "@numberOfReservedSeats";
            numberOfReservedSeatsParameter.Value = numberOfReservedSeats;
            command.Parameters.Add(numberOfReservedSeatsParameter);

            using (var dataReader = command.ExecuteReader())
            {
                if (dataReader.Read())
                {
                    reservation = GetReservationFromResult(dataReader);
                    log.InfoFormat("Exiting Find Reservation with value: " + reservation);
                    return reservation;
                }
            }
        }
        log.InfoFormat("Exiting Find Reservation  with value: " + reservation);
        return reservation;
    }

    
    public Reservation? Save(Reservation reservation)
    {
        log.Info("Saving Reservation");
        IDbConnection connection = DBUtils.getConnection(properties);

        _reservationValidator.Validate(reservation);

        Reservation? reservationFound = FindReservation(reservation.Trip.Id, reservation.ClientName, reservation.ClientContact, reservation.NumberOfReservedSeats);
        if (reservationFound != null)
        {
            log.InfoFormat("Reservation already exists with value: " + reservationFound);
            throw new RepositoryException("Reservation already exists");
        }

        using (var command = connection.CreateCommand())
        {
            command.CommandText =
                "insert into reservation (tripId, clientName, clientContact, numberOfReservedSeats) values (@tripId, @clientName, @clientContact, @numberOfReservedSeats)";
            IDbDataParameter tripParameter = command.CreateParameter();
            tripParameter.ParameterName = "@tripId";
            tripParameter.Value = reservation.Trip.Id;
            command.Parameters.Add(tripParameter);
            
            IDbDataParameter clientNameParameter = command.CreateParameter();
            clientNameParameter.ParameterName = "@clientName";
            clientNameParameter.Value = reservation.ClientName;
            command.Parameters.Add(clientNameParameter);
            
            IDbDataParameter clientContactParameter = command.CreateParameter();
            clientContactParameter.ParameterName = "@clientContact";
            clientContactParameter.Value = reservation.ClientContact;
            command.Parameters.Add(clientContactParameter);
            
            IDbDataParameter numberOfReservedSeatsParameter = command.CreateParameter();
            numberOfReservedSeatsParameter.ParameterName = "@numberOfReservedSeats";
            numberOfReservedSeatsParameter.Value = reservation.NumberOfReservedSeats;
            command.Parameters.Add(numberOfReservedSeatsParameter);

            command.ExecuteNonQuery();

            Reservation? reservationF = FindReservation(reservation.Trip.Id, reservation.ClientName, reservation.ClientContact, reservation.NumberOfReservedSeats);
            reservation.Id = reservationF.Id;
            log.InfoFormat("Reservation saved with value: " + reservation);
        }

        return reservation;
    }

    public Reservation? Delete(int id)
    {
        log.Info("Deleting Reservation");
        IDbConnection connection = DBUtils.getConnection(properties);

        if (id == null)
        {
            throw new RepositoryException("Reservation Id is null");
        }
        
        Reservation? reservationFound = FindOne(id);
        if (reservationFound == null)
        {
            log.InfoFormat("Reservation is not found");
            throw new RepositoryException("Reservation is not found");
        }

        using (var command = connection.CreateCommand())
        {
            command.CommandText = "delete from reservation where id = @id";
            IDbDataParameter parameter = command.CreateParameter();
            parameter.ParameterName = "@id";
            parameter.Value = id;
            command.Parameters.Add(parameter);
        
            command.ExecuteNonQuery();

            log.InfoFormat("Reservation deleted with value: " + reservationFound);
        }

        return reservationFound;
    }

    public Reservation? Update(Reservation reservation)
    {
        log.Info("Updating Reservation");
        IDbConnection connection = DBUtils.getConnection(properties);

        _reservationValidator.Validate(reservation);

        Reservation? reservationFound = FindOne(reservation.Id);
        if (reservationFound == null)
        { 
            log.InfoFormat("Reservation is not found");
            throw new RepositoryException("Reservation is not found");
        }

        using (var command = connection.CreateCommand())
        {
            command.CommandText = "update reservation set tripId = @tripId, numberOfReservedSeats = @numberOfReservedSeats where id = @id";
            IDbDataParameter tripParameter = command.CreateParameter();
            tripParameter.ParameterName = "@tripId";
            tripParameter.Value = reservation.Trip.Id;
            command.Parameters.Add(tripParameter);
            
            IDbDataParameter numberOfReservedSeatsParameter = command.CreateParameter();
            numberOfReservedSeatsParameter.ParameterName = "@numberOfReservedSeats";
            numberOfReservedSeatsParameter.Value = reservation.NumberOfReservedSeats;
            command.Parameters.Add(numberOfReservedSeatsParameter);
            
            IDbDataParameter idParameter = command.CreateParameter();
            idParameter.ParameterName = "@id";
            idParameter.Value = reservation.Id;
            command.Parameters.Add(idParameter);
            
            command.ExecuteNonQuery();
            
            log.InfoFormat("Agency added with value: " + reservation);
        }

        return reservation;
    }

    public IEnumerable<Reservation> FindReservationsByTrip(int tripId)
    {
        log.Info("Finding Reservations By Trip");
        IDbConnection connection = DBUtils.getConnection(properties);
        IList<Reservation> reservations = new List<Reservation>();

        using (var command = connection.CreateCommand())
        {
            command.CommandText = "select " +
                                  "r.id as id, r.numberOfReservedSeats as numberOfReservedSeats, " +
                                  "r.clientName AS clientName, r.clientContact AS clientContact, " +
                                  "t.id as tripId, t.destination as destination, t.transportCompany as transportCompany, " +
                                  "t.departureDate AS departureDate, t.price AS price, t.numberOfAvailableSeats as numberOfAvailableSeats " +
                                  "from reservation r " +
                                  "inner join trip t on t.id = r.tripId " +
                                  "where r.tripId = @tripId";
            IDbDataParameter tripIdParameter = command.CreateParameter();
            tripIdParameter.ParameterName = "@tripId";
            tripIdParameter.Value = tripId;
            command.Parameters.Add(tripIdParameter);
            
            using (var dataReader = command.ExecuteReader())
            {
                while (dataReader.Read())
                {
                    Reservation reservation = GetReservationFromResult(dataReader);
                    reservations.Add(reservation);
                }
            }
        }

        log.InfoFormat("Exiting Find reservations by trip with value: " + reservations);
        return reservations;
    }
    
    public IEnumerable<Reservation> FindReservationsByClient(string clientName)
    {
        log.Info("Finding Reservations By Trip");
        IDbConnection connection = DBUtils.getConnection(properties);
        IList<Reservation> reservations = new List<Reservation>();

        using (var command = connection.CreateCommand())
        {
            command.CommandText = "select " +
                                  "r.id as id, r.numberOfReservedSeats as numberOfReservedSeats, " +
                                  "r.clientName AS clientName, r.clientContact AS clientContact, " +
                                  "t.id as tripId, t.destination as destination, t.transportCompany as transportCompany, " +
                                  "t.departureDate AS departureDate, t.price AS price, t.numberOfAvailableSeats as numberOfAvailableSeats " +
                                  "from reservation r " +
                                  "inner join trip t on t.id = r.tripId " +
                                  "where r.clientName = @clientName";
            IDbDataParameter clientNameParameter = command.CreateParameter();
            clientNameParameter.ParameterName = "@clientName";
            clientNameParameter.Value = clientName;
            command.Parameters.Add(clientNameParameter);
            
            using (var dataReader = command.ExecuteReader())
            {
                while (dataReader.Read())
                {
                    Reservation reservation = GetReservationFromResult(dataReader);
                    reservations.Add(reservation);
                }
            }
        }

        log.InfoFormat("Exiting Find reservations by client with value: " + reservations);
        return reservations;
    }
}